crates/common/download/src/download.rs

```diff
@@ -450,111 +545,84 @@ mod tests {
         }
     }
 
-    // Parameters:
-    //
-    // status code
-    //
-    // bearer token boolean
-    //
-    // maybe url
-    //
-    // expected std error
-    //
-    // description
-    #[test_case(
-        200,
-        false,
-        Some("not_a_url"),
-        "builder error: relative URL without a base"
-        ; "builder error"
-    )]
-    #[test_case(
-        200,
-        true,
-        Some("not_a_url"),
-        "builder error: relative URL without a base"
-        ; "builder error with auth"
-    )]
-    #[test_case(
-        200,
-        false,
-        Some("http://not_a_url"),
-        "dns error: failed to lookup address information"
-        ; "dns error"
-    )]
-    #[test_case(
-        200,
-        true,
-        Some("http://not_a_url"),
-        "dns error: failed to lookup address information"
-        ; "dns error with auth"
-    )]
-    #[test_case(
-        404,
-        false,
-        None,
-        "404 Not Found"
-        ; "client error"
-    )]
-    #[test_case(
-        404,
-        true,
-        None,
-        "404 Not Found"
-        ; "client error with auth"
-    )]
     #[tokio::test]
-    async fn downloader_download_processing_error(
-        status_code: usize,
-        with_token: bool,
-        url: Option<&str>,
-        expected_err: &str,
-    ) -> anyhow::Result<()> {
-        let name = "test_download";
-        let version = Some("test1".to_string());
-        let target_dir_path = TempDir::new()?;
+    async fn resume_download_when_disconnected() -> anyhow::Result<()> {
+        env_logger::init();
+        let chunk_size = 4;
+        let file = "AAAABBBBCCCCDDDD";
+
+        let server_task = tokio::spawn(async move {
```

Comment:
I didn't fully understand this server logic other than the content range manipulation. How does it simulate connection disconnection and resumption? It'd be nice if you could document the overall logic of this server handle.

Reply:
Basically what I was originally going for, was to simulate an HTTP response where a connection just drops and a client hits a timeout. To achieve this, I did this handwritten HTTP server, where it sends only a 1/4th of the response, in one packet, and then drops. But I couldn't find a way to abruptly close the TCP socket without sending a `FIN`: `shutdown()` sends a `FIN`, `Drop` impl does as well, and `mem::forget` doesn't terminate the socket. So I close the socket normally, having only sent one packet with a 1/4th of the response, but the client can detect that it's not the entire response, because it hasn't received the 0 length termination chunk (as we're using `Transfer-Encoding: chunked` here), so while not a complete timeout, it still leads the client to rerequest the missing part of the file.

So you're right, this test definitely needs a comment explaining why it's so complex. I tried to use `mockito`, but I had problems with modifying the contents of the headers. However, now that I have this test complete, I'll once again try to simplify it, hopefully removing that handwritten HTTP logic.

Also, if you have any ideas on how to drop this `TcpStream` here without sending `FIN`, please let me know.